cassiopeia/simulator/Cas9LineageTracingDataSimulator.py

```diff
@@ -0,0 +1,202 @@
+"""
+A Cas9-based lineage tracing data simulator. This is a sublcass of the 
+LineageTracingDataSimulator that simulates the data produced from Cas9-based
+technologies (e.g, as described in Chan et al, Nature 2019 or McKenna et al,
+Science 2016). This simulator implements the method `overlay_data` which takes
+in a CassiopeiaTree with edge lengths and overlays states onto cut-sites.
+"""
+import numpy as np
+
+from cassiopeia.data import CassiopeiaTree
+from cassiopeia.simulator import LineageTracingDataSimulator
+
+
+class Cas9LineageTracingDataSimulator(LineageTracingDataSimulator):
+    """Simulates Cas9-based lineage tracing data.
+
+    This subclass of `LineageTracingDataSimulator` extends the `overlay_data`
+    function to simulate Cas9-based mutations onto a defined "cassette". These
+    cassettes emulate the TargetSites described in Chan et al, Nature 2019 and
+    the GESTALT arrays described in McKenna et al, Science 2020 in which several
+    Cas9 cut-sites are arrayed together. In the Chan et al technology, these
+    cassettes are of length 3; in McKenna et al, these cassettes are of length
+    10.
+
+    The class accepts several parameters that govern the Cas9-based tracer.
+    First and foremost is the cutting rate, describing how fast Cas9 is able to
+    cut. We model Cas9 cutting as an exponential process, parameterized by the
+    specified mutation rate - specifically, for a lifetime `t` and parameter
+    `lambda`, the expected probability of Cas9 mutation, per site, is
+    `exp(-lambda * t)`.
+    
+    Second, the class accepts architecture of the recorder - described by the
+    size of the cassette (by default 3) and the number of cassettes. The
+    resulting lineage will have |cassette|*(# of cassettes) characters.
+
+    Third, the class accepts a state distribution describing the relative
+    likelihoods of various indels. This is very useful, as it is typical that a
+    handful of mutations are far likelier than the bulk of the possible
+    mutations.
+
+    Finally, the class accepts a silencing rate. This is normally a rare event
+    in which an entire cassette is transcriptionally silenced and therefore
+    not observed.
+
+    The function `overlay_data` will operate on the tree in place and will
+    specifically modify the data stored in the character attributes.
+
+    Args:
+        number_of_cassettes: Number of cassettes (i.e., arrays of target sites)
+        size_of_cassette: Number of editable target sites per cassette
+        mutation_rate: Exponential parameter for the Cas9 cutting rate.
```

Comment:
Have you considered having a different mutation rate for each site?

Reply:
We have done this experiment before -- you can see it in the "Phased Editor" part of our Genome Biology paper. We could easily add this to the simulator.